includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -1285,61 +1298,112 @@ public function get_related_products_query( $cats_array, $tags_array, $exclude_i
+			$stock_quantity,
+			$product_id_with_stock
+		);
+
+		$sql = apply_filters( 'woocommerce_update_product_stock_query', $sql, $product_id_with_stock, $stock_quantity, 'set' );
+
+		$wpdb->query( $sql ); // phpcs:ignore WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.PreparedSQL.NotPrepared
+
+		// Cache delete is required (not only) to set correct data for lookup table (which reads from cache).
+		// Sometimes I wonder if it shouldn't be part of update_lookup_table.
+		wp_cache_delete( $product_id_with_stock, 'post_meta' );
+
+		// Stock qty in lookup table might not be set to the same value here due to manage_stock not being set to true, thus using force update.
+		$this->update_lookup_table( $product_id_with_stock, 'wc_product_meta_lookup', true );
+	}
+
 	/**
 	 * Update a product's stock amount directly.
 	 *
 	 * Uses queries rather than update_post_meta so we can do this in one query (to avoid stock issues).
+	 * Ignores manage stock setting on the product and sets quantities directly in the db: post meta and lookup tables.
+	 * Uses locking to update the quantity. If the lock is not acquired, change is lost.
 	 *
 	 * @since  3.0.0 this supports set, increase and decrease.
-	 * @param  int      $product_id_with_stock Product ID.
-	 * @param  int|null $stock_quantity Stock quantity.
-	 * @param  string   $operation Set, increase and decrease.
+	 * @param  int            $product_id_with_stock Product ID.
+	 * @param  int|float|null $stock_quantity Stock quantity.
+	 * @param  string         $operation Set, increase and decrease.
+	 * @return int|float New stock level.
 	 */
 	public function update_product_stock( $product_id_with_stock, $stock_quantity = null, $operation = 'set' ) {
 		global $wpdb;
+
+		// Ensures a row exists to update.
 		add_post_meta( $product_id_with_stock, '_stock', 0, true );
 
-		// Update stock in DB directly.
-		switch ( $operation ) {
-			case 'increase':
-				$sql = $wpdb->prepare(
-					"UPDATE {$wpdb->postmeta} SET meta_value = meta_value + %f WHERE post_id = %d AND meta_key='_stock'",
-					$stock_quantity,
-					$product_id_with_stock
-				);
-				break;
-			case 'decrease':
-				$sql = $wpdb->prepare(
-					"UPDATE {$wpdb->postmeta} SET meta_value = meta_value - %f WHERE post_id = %d AND meta_key='_stock'",
-					$stock_quantity,
-					$product_id_with_stock
-				);
-				break;
-			default:
-				$sql = $wpdb->prepare(
-					"UPDATE {$wpdb->postmeta} SET meta_value = %f WHERE post_id = %d AND meta_key='_stock'",
-					$stock_quantity,
-					$product_id_with_stock
+		if ( 'set' === $operation ) {
+			// 'set' does not need locking as it directly updates db to the new value.
+			$new_stock = wc_stock_amount( $stock_quantity );
+			$this->set_product_stock( $product_id_with_stock, $new_stock );
+		} else {
+			// 'fetch-and-update' atomic operation for stock quantity.
+			try {
+				wc_transaction_query( 'start' );
```

Comment:
I believe it would be prudent to get feedback from someone with more experience with databases before proceeding with this approach. In the past, we had to revert commits that used transactions due to unexpected issues (#18175 and #20537).

My concern here is that there is some application time between the `SELECT FOR UPDATE` and the `UPDATE` that is executed in `WC_Product_Data_Store_CPT::set_product_stock()` and that could cause the locks on the rows to block other clients who are running the same queries. I imagine this could be a problem for busy stores, as I think this transaction is executed whenever an order that includes a stock managed product is placed.

Reply:
> My concern here is that there is some application time between the SELECT FOR UPDATE and the UPDATE that is executed in WC_Product_Data_Store_CPT::set_product_stock() and that could cause the locks on the rows to block other clients who are running the same queries.
Well, that's the purpose of a lock, so I would hope so. I was hoping to make the query non blocking by using NOWAIT, but seems like that's not an option. 

In addition, reading through [MySQL 5.5 manual](https://dev.mysql.com/doc/refman/5.5/en/innodb-deadlock-detection.html), I can see that InnoDB has some other issues as well:
> When InnoDB performs a complete rollback of a transaction, all locks set by the transaction are released. However, if just a single SQL statement is rolled back as a result of an error, some of the locks set by the statement may be preserved. This happens because InnoDB stores row locks in a format such that it cannot know afterward which lock was set by which statement.

I've removed the locks in af79509.